lib/ansible/modules/windows/win_format.ps1

```diff
@@ -0,0 +1,157 @@
+    $module.FailJson("The parameter drive_letter should be a single character A-Z")
+}
+if ($integrity_streams -eq $true -and $file_system -ne "refs") {
+    $module.FailJson("Integrity streams can be enabled only on ReFS volumes. You specified: $($file_system)")
+}
+if ($compress_volume -eq $true) {
+    if ($file_system -eq "ntfs") {
+        if ($null -ne $allocation_unit_size -and $allocation_unit_size -gt 4096) {
+            $module.FailJson("NTFS compression is not supported for allocation unit sizes above 4096")
+        }
+    }
+    else {
+        $module.FailJson("Compression can be enabled only on NTFS volumes. You specified: $($file_system)")
+    }
+}
+
+function Get-AnsibleVolume {
+    param(
+        $DriveLetter,
+        $Path,
+        $Label
+    )
+
+    if ($null -ne $DriveLetter) {
+        try {
+            $volume = Get-Volume -DriveLetter $DriveLetter
+        } catch {
+            $module.FailJson("There was an error retrieving the volume using drive_letter $($DriveLetter): $($_.Exception.Message)", $_)
+        }
+    }
+    elseif ($null -ne $Path) {
+        try {
+            $volume = Get-Volume -Path $Path
+        } catch {
+            $module.FailJson("There was an error retrieving the volume using path $($Path): $($_.Exception.Message)", $_)
+        }
+    }
+    elseif ($null -ne $Label) {
+        try {
+            $volume = Get-Volume -FileSystemLabel $Label
+        } catch {
+            $module.FailJson("There was an error retrieving the volume using label $($Label): $($_.Exception.Message)", $_)
+        }
+    }
+    else {
+        $module.FailJson("Unable to locate volume: drive_letter, path and label were not specified")
+    }
+
+    return $volume
+}
+
+function Get-VolumeParams {
+    param(
+        $Path,
+        $Label,
+        $FileSystem,
+        $Full,
+        $UseLargeFRS,
+        $Compress,
+        $SetIntegrityStreams
+    )
+    $parameters = @{
+        Path = $Path
+        Full = $Full
+    }
+    if ($null -ne $UseLargeFRS) {
+        $parameters.Add("UseLargeFRS", $UseLargeFRS)
+    }
+    if ($null -ne $SetIntegrityStreams) {
+        $parameters.Add("SetIntegrityStreams", $SetIntegrityStreams)
+    }
+    if ($null -ne $compress_volume){
+        $parameters.Add("Compress", $Compress)
+    }
+    if ($null -ne $Label) {
+        $parameters.Add("NewFileSystemLabel", $Label)
+    }
+    if ($null -ne $FileSystem) {
+        $parameters.Add("FileSystem", $FileSystem)
+    }
+
+    return $parameters
+}
+
+$ansible_volume = Get-AnsibleVolume -DriveLetter $drive_letter -Path $path -Label $label
+
+if (-not $force_format) {
+    $ansible_volume_size = $ansible_volume.Size
+    $ansible_partition_size = (Get-Partition -Volume $ansible_volume).Size
+
+    if ($ansible_volume_size -ne 0 -or $ansible_partition_size -eq $ansible_volume_size) {
+        $module.FailJson("You must specify force as a parameter to format non-pristine volumes")
```

Comment:
Windows automatically picks up an appropriate file system if it's not specified so it's a required parameter. Default file systems can vary depending on the device and target OS so I wasn't sure how to deal with this.

I suppose I can add a check to implement this only if the file system was explicitly specified.

Reply:
If the file system is optional then I think that adds more ammunition that this check should be omitted when not specified. It means that the user only cared that the volume was formatted but not the details and so a non-raw volume meets their criteria. They can always set `force_format` to always format the volume but they want the task to say OK with no change otherwise.

I think we can also say the same thing with an explicit filesystem if they already match. Maybe with a `strict` param or something that causes it to fail if the file system matches but some of the extra details like allocation size and so on. We already have examples of some modules where resources are created and subsequent runs ignore checking the extra options.

Reply:
I would very much like it if I haven't changed any module parameter values and I re-run I just get 'OK'.  If there are some parameters that can't be changed except by reformatting then as long as it says so in the docs and you have to use `force_format`: to reformat it, then I'd be happy with this.

Otherwise this is great and too useful to be without, I've already added it some of my test playbooks.